ouroboros-consensus/src/Ouroboros/Consensus/Ledger/Query.hs

```diff
@@ -16,27 +26,142 @@ import           Data.Maybe (isJust)
 
+-- | Different queries supported by the ledger for all block types, indexed
+-- by the result type.
+data Query blk result = BlockQuery (BlockQuery blk result)
+
+instance (ShowProxy (BlockQuery blk)) => ShowProxy (Query blk) where
+  showProxy (Proxy :: Proxy (Query blk)) = "Query (" ++ showProxy (Proxy @(BlockQuery blk)) ++ ")"
+
+instance (ShowQuery (BlockQuery blk)) => ShowQuery (Query blk) where
+  showResult (BlockQuery blockQuery) = showResult blockQuery
+
+instance Eq (SomeSecond BlockQuery blk) => Eq (SomeSecond Query blk) where
+  SomeSecond (BlockQuery blockQueryA) == SomeSecond (BlockQuery blockQueryB)
+    = SomeSecond blockQueryA == SomeSecond blockQueryB
+
+instance Show (SomeSecond BlockQuery blk) => Show (SomeSecond Query blk) where
+  show (SomeSecond (BlockQuery blockQueryA)) = "Query " ++ show (SomeSecond blockQueryA)
+
+queryEncodeNodeToClient ::
+    forall blk. (
+      SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+    )
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> (SomeSecond Query blk)
+  -> Encoding
+queryEncodeNodeToClient codecConfig version blockVersion (SomeSecond query)
+  | not (version >= NodeToClientV_9)
+  = error "encode failure: Serializing Query is not supported"
+  | otherwise
+  = case query of
+    BlockQuery blockQuery ->
+      encodeTag 0 <> encodeNodeToClient
+                      @blk
+                      @(SomeSecond BlockQuery blk)
+                      codecConfig
+                      blockVersion
+                      (SomeSecond blockQuery)
+
+queryDecodeNodeToClient ::
+    forall blk. (
+      SerialiseNodeToClient blk (SomeSecond BlockQuery blk)
+    )
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> forall s. Decoder s (SomeSecond Query blk)
+queryDecodeNodeToClient codecConfig version blockVersion
+  | not (version >= NodeToClientV_9)
+  = fail $ "decode failure: Deserializing Query is not supported (on node to client version: " ++ show version ++ ")"
+  | otherwise
+  = do
+    tag <- decodeTag
+    case tag of
+      0 -> do
+        SomeSecond x <- decodeNodeToClient
+            @blk
+            @(SomeSecond BlockQuery blk)
+            codecConfig
+            blockVersion
+        return (SomeSecond (BlockQuery x))
+      _ -> fail $ "SomeSecond Query blk: unknown tag " ++ show tag
+
+queryEncodeResult ::
+     forall blk result. SerialiseResult blk (BlockQuery blk)
+  => CodecConfig blk
+  -> NodeToClientVersion
+  -> BlockNodeToClientVersion blk
+  -> Query blk result
+  -> result
+  -> Encoding
+queryEncodeResult codecConfig version blockNodeToClientVersion query result
+  | not (version >= NodeToClientV_9)
+  = error $ "encode failure: Serializing Query result is not supported (on node to client version: " ++ show version ++ ")"
+  | otherwise
+  = case query of
+      BlockQuery blockQuery ->
+        encodeTag 0
```

Comment:
My other comment suggests we put this logic off to the next PR, after merging this one. That said, I don't think the results will ever need an additional tag; the query is argument to `encodeResult` only so that we know which type we're serializing.